clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,258 @@
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  Clock write -> Clock read -> Reset read ->
+
+  -- | Write data
+  Signal write (BitVector n) ->
+  -- | Write enable
+  Signal write Bool ->
+  -- | Read enable
+  Signal read Bool ->
+  ( Signal write ResetBusy
+  , Signal write Full
+  , Signal write (DataCount depth)
+
+  , Signal read ResetBusy
+  , Signal read Empty
+  , Signal read (DataCount depth)
+  , Signal read (BitVector n)
+  )
+dcFifo# DcConfig{..} wClk rClk rst writeData wEnable rEnable =
+  let
+    (wRstBusy, wFull, wCnt, rRstBusy, rEmpty, rCnt, rData) = go rstSignalR rstSignalW start wEnable rEnable writeData
+  in
+    ( wRstBusy
+    , wFull
+    , if dcWriteDataCount then wCnt else deepErrorX "Write data count disabled"
+    , rRstBusy
+    , rEmpty
+    , if dcReadDataCount then rCnt else deepErrorX "Read data count disabled"
+    , deepErrorX "No sample" :- rData
+    )
+ where
+  rstSignalR = unsafeToHighPolarity rst
+  rstSignalW = unsafeSynchronizer rClk wClk $ unsafeToHighPolarity rst
+
+  -- reified depth
+  rD = snatToNum @Int (powSNat (SNat @2) (SNat @depth))
+
+  -- TODO: how do we ensure it doesn't "space leak" in Haskell?
+  --
+  -- force slower first?
+  -- https://github.com/clash-lang/clash-compiler/blob/ea114d8edd6a110f72d148203b9db2454cae8f37/clash-prelude/src/Clash/Explicit/BlockRam.hs#L1276-L1280
+
+  go ::
+    Signal read Bool ->
+    Signal write Bool ->
+
+    FifoState n ->
+    Signal write Bool ->
+    Signal read Bool ->
+    Signal write (BitVector n) ->
+    ( Signal write ResetBusy
+    , Signal write Full
+    , Signal write (DataCount depth)
+
+    , Signal read ResetBusy
+    , Signal read Empty
+    , Signal read (DataCount depth)
+    , Signal read (BitVector n)
+    )
+  go rstR rstW st@(FifoState _ rt) wEna rEna =
+    if rt < tWr
+      then goRead rstR rstW st wEna rEna
+      else goWrite rstR rstW st wEna rEna
+    -- TODO: goBoth case?
+
+  goWrite rstR (True :- rstWNext) (FifoState _ rt) (_ :- wEna) rEna (_ :- wData) =
+      (1 :- wRstBusy, 0 :- preFull, 0 :- preWCnt, rRstBusy, fifoEmpty, rCnt, rData)
+    where
+      (wRstBusy, preFull, preWCnt, rRstBusy, fifoEmpty, rCnt, rData) = go rstR rstWNext (FifoState mempty (rt-tWr)) wEna rEna wData
+
+  goWrite rstR (_ :- rstW) (FifoState q rt) wEna rEna wData =
+    (0 :- wRstBusy, full, wCnt, rRstBusy, fifoEmpty, rCnt, rData)
+    where
+      (wRstBusy, preFull, preWCnt, rRstBusy, fifoEmpty, rCnt, rData) = go rstR rstW (FifoState q' (rt-tWr)) wEna' rEna wData'
+      wCnt = sDepth q :- preWCnt
+      full = (if Seq.length q == rD then high else low) :- preFull
+      (en :- wEna') = wEna
+      (wDatum :- wData') = wData
+      q' =
+        if Seq.length q + 1 <= rD && en
+          then wDatum Seq.<| q
+          else q
+
+  sDepth = fromIntegral . Seq.length
+
+  goRead (rstR :- rstRNext) rstW (FifoState q rt) wEna rEna wData =
+    (wRstBusy, full, wCnt, (if rstR then 1 else 0) :- rRstBusy, fifoEmpty, rCnt, rData)
```

Comment:
If you'd use `bitToBool` in the blackbox, you could do:

```suggestion
    (wRstBusy, full, wCnt, rstR :- rRstBusy, fifoEmpty, rCnt, rData)
```

which also makes me think we can just move it out of the return type entirely.

Reply:
Yes! I think I have it working.